pkg/extensions/tbtc/tbtc.go

```diff
@@ -0,0 +1,286 @@
+
+	return nil
+}
+
+// TODO:
+//  1. Filter incoming events by operator interest.
+//  2. Incoming events deduplication.
+//  3. Resume monitoring after client restart.
+func (t *tbtc) monitorAndAct(
+	ctx context.Context,
+	monitoringName string,
+	monitoringStartFn watchDepositEventFn,
+	monitoringStopFn watchDepositEventFn,
+	keepClosedFn watchKeepClosedFn,
+	actFn submitDepositTxFn,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) (subscription.EventSubscription, error) {
+	handleStartEvent := func(deposit string) {
+		logger.Infof(
+			"starting [%v] monitoring for deposit [%v]",
+			monitoringName,
+			deposit,
+		)
+
+		stopEventChan := make(chan struct{})
+
+		stopEventSubscription, err := monitoringStopFn(
+			func(stopEventDeposit string) {
+				if deposit == stopEventDeposit {
+					stopEventChan <- struct{}{}
+				}
+			},
+		)
+		if err != nil {
+			logger.Errorf(
+				"could not setup stop event handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer stopEventSubscription.Unsubscribe()
+
+		keepClosedChan, keepClosedUnsubscribe, err := keepClosedFn(deposit)
+		if err != nil {
+			logger.Errorf(
+				"could not setup keep closed handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer keepClosedUnsubscribe()
+
+		timeoutChan := time.After(timeout)
+
+		actionAttempt := 1
+
+	monitoring:
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Infof(
+					"context is done for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-stopEventChan:
+				logger.Infof(
+					"stop event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-keepClosedChan:
+				logger.Infof(
+					"keep closed event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-timeoutChan:
+				err := actFn(deposit)
+				if err != nil {
+					if actionAttempt == maxActionAttempts {
+						logger.Errorf(
+							"could not perform action "+
+								"for [%v] monitoring for deposit [%v]: [%v]; "+
+								"last attempt failed",
+							monitoringName,
+							deposit,
+							err,
+						)
+						break monitoring
+					}
+
+					backoff := actBackoffFn(actionAttempt)
+
+					logger.Errorf(
+						"could not perform action "+
+							"for [%v] monitoring for deposit [%v]: [%v]; "+
+							"retrying after: [%v]",
+						monitoringName,
+						deposit,
+						err,
+						backoff,
+					)
+
+					timeoutChan = time.After(backoff)
+					actionAttempt++
+				} else {
+					logger.Infof(
+						"action for [%v] monitoring for "+
+							"deposit [%v] has been performed successfully",
+						monitoringName,
+						deposit,
+					)
```

Comment:
I would rather log that we are going to do something (my comment above) and then log error in case the action failed instead of logging this. We are not sure if the action succeeded - we submitted it to the chain, it's still waiting to be mined and there is still a small chance something can co wrong.